src/modules/auth/google-oauth/google-oauth.controller.ts

```diff
@@ -0,0 +1,34 @@
+import { Controller, Get, Redirect, Req, UseGuards } from '@nestjs/common';
+import { Request } from 'express';
+import { GoogleOauthService } from 'src/modules/auth/google-oauth/google-oauth.service';
+import { GoogleOauthSessionGuard } from 'src/modules/auth/google-oauth/guards/google-oauth-session.guard';
+import { GoogleOauthGuard } from 'src/modules/auth/google-oauth/guards/google-oauth.guard';
+import { GoogleUser } from 'src/modules/auth/google-oauth/strategies/google-oauth.strategy';
+
+export interface GoogleReq extends Request {
+  user: GoogleUser;
+}
+
+@Controller('auth/google')
+export class GoogleOauthController {
+  constructor(private googleOauthService: GoogleOauthService) {}
+
+  @Get()
+  @UseGuards(GoogleOauthGuard)
+  async googleAuth() {
+    // Guard redirects
+  }
+
+  @Get('redirect')
```

Comment:
will google use this endpoint or client(react app)?

Reply:
This one Google will use after the user signs in through their service. The client will use /auth/google in order to go through Google's authentication and then after the user clicks the sign-in button, Google will redirect the browser to this route (auth/google/redirect). This route then will sign up the user if he isn't already and then redirect him back to the Frontend.

I've used the Redirect() decorator to redirect after all this process. If you want I can also refactor like this:
```typescript
  @Get('redirect')
  @UseGuards(GoogleOauthGuard)
  async googleAuthRedirect(@Req() req: GoogleReq, @Res() res: Response) {
    this.googleOauthService.googleSignUp(req);
    return res.redirect(process.env.FRONTEND_REDIRECT_URL, 301)
  }
```